src/impls/vec.rs

```diff
@@ -1,20 +1,22 @@
 use crate::prelude::*;
 
 impl<T: StableHash> StableHash for Vec<T> {
-    fn stable_hash<H: StableHasher>(&self, sequence_number: H::Seq, state: &mut H) {
+    fn stable_hash<H: StableHasher>(&self, field_address: H::Addr, state: &mut H) {
         profile_method!(stable_hash);
 
-        (&self[..]).stable_hash(sequence_number, state)
+        (&self[..]).stable_hash(field_address, state)
     }
 }
 
 impl<'a, T: StableHash> StableHash for &'a [T] {
-    fn stable_hash<H: StableHasher>(&self, mut sequence_number: H::Seq, state: &mut H) {
+    fn stable_hash<H: StableHasher>(&self, field_address: H::Addr, state: &mut H) {
         profile_method!(stable_hash);
 
-        for item in self.iter() {
-            item.stable_hash(sequence_number.next_child(), state);
+        for (index, item) in self.iter().enumerate() {
+            item.stable_hash(field_address.child(index as u64), state);
         }
-        self.len().stable_hash(sequence_number, state);
+        // This is needed to disambiguate when the last members are default
```

Comment:
I don't (fully) understand this comment - is this so we can distinguish, e.g., between a `Vec` with one entry `true`, and a `Vec` where the first entry is `true` followed by any number of `false`? Would be good if the comment spelled that out a little more (and the crate docs should mention that vecs of different length will have different hashes, even if they are filled with inert values)

Reply:
You are correct that this disambiguates between eg: `vec![true, false]` and `vec![true, false, false]` or `vec![true]`. (The test pointed to in the "See also" is an example of this, but using ints). It does not seem particularly "surprising" enough to be called out that vecs of different length will have different hashes even if filled with inert values _All_ possible values for _any_ possible schema do not collide within the same schema. That is, when holding the schema constant, the encoding is injective. (If you find this not to be true somewhere it's a show-stopping bug)